shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,522 @@
+//! Module emulating the pow function as found in the Balancer contract code.
+// The original contract code can be found at:
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/LogExpMath.sol
+
+#![allow(dead_code)]
+
+use ethcontract::{I256, U256};
+use lazy_static::lazy_static;
+use std::convert::TryFrom;
+
+lazy_static! {
+    static ref ONE_18: I256 = I256::exp10(18);
+    static ref ONE_20: I256 = I256::exp10(20);
+    static ref ONE_36: I256 = I256::exp10(36);
+    static ref UNSIGNED_ONE_18: U256 = U256::try_from(*ONE_18).unwrap();
+    static ref MAX_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(130_i128)).unwrap();
+    static ref MIN_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(-41_i128)).unwrap();
+    static ref LN_36_LOWER_BOUND: I256 = ONE_18.checked_sub(I256::exp10(17)).unwrap();
+    static ref LN_36_UPPER_BOUND: I256 = ONE_18.checked_add(I256::exp10(17)).unwrap();
+    static ref MILD_EXPONENT_BOUND: U256 = (U256::one() << 254_u32)
+        .checked_div(U256::try_from(*ONE_20).unwrap())
+        .unwrap();
+}
+fn constant_x_20(i: u32) -> I256 {
```

Comment:
I wonder if there is a cleaver way to make these const-fns... (not actually suggesting the change, just mildly curious :stuck_out_tongue:).